op/reboot.go

```diff
@@ -0,0 +1,394 @@
+		}
+		n.Spec.Unschedulable = c.unschedulable
+		newData, err := json.Marshal(n)
+		if err != nil {
+			return err
+		}
+		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, n)
+		if err != nil {
+			return fmt.Errorf("failed to create patch for node %s: %v", n.Name, err)
+		}
+		_, err = nodesAPI.Patch(ctx, n.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to patch node %s: %v", n.Name, err)
+		}
+	}
+	return nil
+}
+
+func (c cordonCommand) Command() cke.Command {
+	ipAddresses := make([]string, len(c.nodes))
+	for i, n := range c.nodes {
+		ipAddresses[i] = n.Address
+	}
+	return cke.Command{
+		Name:   "cordonCommand",
+		Target: strings.Join(ipAddresses, ","),
+	}
+}
+
+type drainCommand struct {
+	timeoutSeconds      int
+	apiserver           *cke.Node
+	nodes               []*cke.Node
+	protectedNamespaces *metav1.LabelSelector
+}
+
+func listNamespacesForDrain(ctx context.Context, cs *kubernetes.Clientset, ls *metav1.LabelSelector) (map[string]bool, error) {
+	all, err := cs.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(ls)
+	if err != nil {
+		return nil, err
+	}
+	protected, err := cs.CoreV1().Namespaces().List(ctx, metav1.ListOptions{
+		LabelSelector: selector.String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	nss := make(map[string]bool)
+	for _, ns := range all.Items {
+		nss[ns.Name] = false
+	}
+	for _, ns := range protected.Items {
+		nss[ns.Name] = true
+	}
+
+	return nss, nil
+}
+
+func evictOrDeleteNodePod(ctx context.Context, cs *kubernetes.Clientset, n *cke.Node, namespaces map[string]bool) ([]*corev1.Pod, error) {
+	var targets []*corev1.Pod
+	for ns, protected := range namespaces {
+		podsAPI := cs.CoreV1().Pods(ns)
+		podList, err := podsAPI.List(ctx, metav1.ListOptions{
+			FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": n.Nodename()}).String(),
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		for _, pod := range podList.Items {
+			owner := metav1.GetControllerOf(&pod)
+			if owner != nil && owner.APIVersion == "apps/v1" && owner.Kind == "DaemonSet" {
+				continue
+			}
+			targets = append(targets, &pod)
+			err := podsAPI.Evict(ctx, &v1beta1.Eviction{
+				TypeMeta:      metav1.TypeMeta{Kind: "Pod", APIVersion: "v1"},
+				ObjectMeta:    metav1.ObjectMeta{Name: pod.Name, Namespace: pod.Namespace},
+				DeleteOptions: &metav1.DeleteOptions{},
+			})
+			switch {
+			case apierrors.IsNotFound(err):
+			case apierrors.IsTooManyRequests(err) && !protected:
+				err := podsAPI.Delete(ctx, pod.Name, metav1.DeleteOptions{})
+				if err != nil {
+					return nil, err
+				}
+			case err != nil:
+				return nil, err
+			}
+		}
+	}
+	return targets, nil
+}
+
+func waitPodDeletion(ctx context.Context, cs *kubernetes.Clientset, pods []*corev1.Pod, timeoutSeconds int) error {
+	if timeoutSeconds != 0 {
```

Comment:
I think waiting forever for pod deletion is not good.
It would block other operations of CKE to keep the cluster healthy.